Add Port-Forwarding and UPNP tutorial

TheYellowArchitect opened the pull request:
There are no Universal Plug n Play text tutorials on the internet. I only found these 2 videos:
1. https://www.youtube.com/watch?v=d8QpnamQq1A2
2. The finale of https://www.youtube.com/watch?v=nQ4P3ogXp2Q

Instead of a user searching blindly on youtube, a basic UPNP tutorial should be on the Godot docs.
I myself used this tutorial with 2 different routers, and can confirm it works.

As a necessary side-note, UPNP is very important because its how you test with friends, before deploying on a real server. And as seen in this text tutorial, its a 10 minute task, and you can play with anyone.

For convenience (since .html comes with a lot of baggage), I upload how the final version looks in images:
![image1](https://user-images.githubusercontent.com/25630803/228348075-e3fd489a-7253-4f19-a4f6-96b7f5d53b8d.png)
![image2](https://user-images.githubusercontent.com/25630803/228348090-a0924471-f8eb-4bae-8ae6-17f6d2d83caf.png)
![image3](https://user-images.githubusercontent.com/25630803/228348105-3ea2a0a8-b8ee-4f36-9539-3e57ff0dd878.png)
![image4](https://user-images.githubusercontent.com/25630803/228348126-c183023f-9e1f-4351-b6fa-56b23604ba69.png)
![image5](https://user-images.githubusercontent.com/25630803/228348141-a9a0c538-80e5-4246-88c4-3c1a251118aa.png)
![image6](https://user-images.githubusercontent.com/25630803/228348154-5256011a-88bc-451a-9f54-5941fdd507a5.png)

Feel free to overwrite anything, all I want is more devs who are currently netcoding, to learn of UPNP simply because of how convenient it is. Basically localhost but your friends can connect to you.

mhilbrunner commented on tutorials/networking/port_forwarding_upnp.rst:
```diff
@@ -0,0 +1,109 @@
+        peer.create_client(ip, port)
+        multiplayer.multiplayer_peer = peer
+
+and convert it into UPNP. Reminder that **only the host must use UPNP**. Because the goal is after all to expose a public IP onto the wide internet, accessible by anyone.
```
```suggestion
and convert it so it uses UPnP. Reminder that **only the host needs UPnP**. Because the goal is after all to expose the host's port to the wide internet, accessible by anyone.
```

mhilbrunner reviewed (commented):
Hey, thanks for working on this! Creating docs for UPnP has been on my to do list for quite some time since I originally added UPnP support to Godt, but I always lacked the time.

I started on this PR but had quite a bit of feedback before I ran out of time. Maybe it's of use to you already.

Also, we should choose one spelling of **Universal Plug and Play** and **UPnP** and keep them everywhere on this page. I suggest these ones, they seem to be used by Wikipedia and official UPnP material.

TheYellowArchitect commented on tutorials/networking/port_forwarding_upnp.rst:
```diff
@@ -0,0 +1,109 @@
+---------------------------
+
+After you are done testing solo with localhost(127.0.0.1) you want to test on a *real* connection.
+You will be likely overwhelmed with your choices:
+
+1. Host a real server and deploy there (deployment of a build will be a chore unless you learn advanced methods like remote ssh deployment which are included in Godot)
+
+2. Peer-to-Peer connection, straight to a computer. This includes NAT Holepunching, STUN/TURN and other methods which are overwhelming compared to :abbr:`UPNP (Universal Plug n Play)`.
+
+Before any code, let's look into the router setup; Not all Routers support :abbr:`UPNP (Universal Plug n Play)`, nor do all users have :abbr:`UPNP (Universal Plug n Play)` enabled. For this reason it is suggested you use :abbr:`UPNP (Universal Plug n Play)` for testing instead of your final public deployment. Even if you enable your router's :abbr:`UPNP (Universal Plug n Play)` (assuming it is a  ...
```
>Also the real possibility of just hosting locally and forwarding the port manually once (and maybe using DynDNS to deal with changing IPs) should be mentioned, as its probably the easiest way (or hosting yourself if you know what you're doing).

I tried just opening port 9999 (basically the ruleset of this tutorial) and hosting on my public IP, but I couldn't have (local) clients connect to my external public IP. Unless I hosted via UPNP. No idea if its a bug or I did something wrong (probably the latter) but if you can host locally and have others connect to you from a non-local network without UPNP, do share the secret :P

mhilbrunner replied:
You likely need to setup port forwarding on your router to forward the external port 9999 on your public IP to port 9999 on your local internal machine. Basically what UPnP does for you, manually. Not too much of a burden if you only have to do it once for your own game, its a few clicks and doesn't require enabling or messing around with UPnP or adding code for that.

UPnP is (was) mostly useful to do this automatically for less technical users, but as more and more routers come with UPnP disabled due to security concerns, it's growing less and less useful - if you need to teach users how to enable UPnP (if possible at all), you may as well just tell them to create a port forward in the router manually.

TheYellowArchitect commented on tutorials/networking/port_forwarding_upnp.rst:
```diff
@@ -0,0 +1,109 @@
+How to port-forward
+-------------------
+
+1. Login to your router (username/password is on the router's bottom where it has barcode tag and original wifi password)
```
>We should not provide guidance on how to login to your router. That is way too specific, depends on model and vendor and is also just not our scope. :)

I browsed UPNP setups a bit before I found the 2 videos mentioned in OP whereas I got the final UPNP solution mentioned here. Many people who failed (e.g. https://teddit.net/r/godot/comments/121lul7/godot_4_upnp_invalid_gateway), I bet just failed the port forwarding process. 

As for what you said (quoted above), I tested this succesfully on 2 routers. 1 of them did not have UPNP at all, I just port forwarded and it worked with the UPNP code of this tutorial. And by extension, the sub-tutorial of 'How to port forward' applies to basically all routers, and ensures the user will open his ports instead of trying to just enable his router's UPNP, or be discouraged if he finds his router doesn't have UPNP.

>We should just say something to the effect of "make sure UPnP port forwarding is enabled and supported by your router"

Ideally it would be a collapsable section. I don't know how to do that, does sphinx support such an option?

>Also, I don't get the explanation on creating a port-forward; UPnP is used exactly because it avoids having to do that.

Mentioned far above, tl;dr: Enabling UPnP without opening a port isn't enough for 2/2 routers I used, for external clients to connect.

TheYellowArchitect replied:
>If you open and forward a port in your router manually, you don't need UPnP. It will just work. 

It didn't in my case. I will make an MRP when I try again, though that will take a while as it is low priority~

TheYellowArchitect commented:
To find a full review not even a day after was unexpected. I thank you for the swift response, and of course the feedback itself. I incorporated all of it, and added a bit of polish.

@mhilbrunner

skyace65 commented on tutorials/networking/port_forwarding_upnp.rst:
```diff
@@ -0,0 +1,112 @@
+
+Besides testing locally on your own, you should test on a *real* network connection to make sure your game actually works under real-life networking conditions. There are multiple ways to achieve this:
+
+    1. Keep hosting your game locally, but expose it to the public internet. This is easy and quick, but comes with a few downsides. Your public IP is likely dynamically assigned and changes, which requires you to let your testers know each time what to connect to, or to use a dynamic DNS service to point at your changing IP. You also need to deal with any firewalls blocking external access to yo ...
```
```suggestion
    1. Keep hosting your game locally, but expose it to the public internet. This is easy and quick, but comes with a few downsides. Your public IP is likely dynamically assigned and changes, which requires you to let your testers know each time what to connect to, or to use a dynamic DNS service to point at your changing IP. You also need to deal with any firewalls blocking external access to your local game, most notably on your router, and in the likely case it uses NAT (network address translation), you also need to configure port forwarding on your router. Finally, you open a (small part of) your computer and home network to the public internet, which always has some risks.
```
